File: templates/types/streaming/fastapi/app/engine/query_filter.py
from llama_index.core.vector_stores.types import MetadataFilter, MetadataFilters


def generate_filters(doc_ids):
    """
    Generate public/private document filters based on the doc_ids and the vector store.
    """
    public_doc_filter = MetadataFilter(
        key="private",
        value="true",
        operator="!=",  # type: ignore
    )
    selected_doc_filter = MetadataFilter(
        key="doc_id",
        value=doc_ids,
        operator="in",  # type: ignore
    )
    if len(doc_ids) > 0:
        # If doc_ids are provided, we will select both public and selected documents
        filters = MetadataFilters(
            filters=[
                public_doc_filter,
                selected_doc_filter,
            ],
            condition="or",  # type: ignore
        )
    else:
        filters = MetadataFilters(
            filters=[
                public_doc_filter,
            ]
        )

    return filters

File: packages/create-llama/templates/components/vectordbs/typescript/llamacloud/queryFilter.ts
import { CloudRetrieveParams, MetadataFilter } from "llamaindex";

export function generateFilters(documentIds: string[]) {
  // public documents (ingested by "npm run generate" or in the LlamaCloud UI) don't have the "private" field
  const publicDocumentsFilter: MetadataFilter = {
    key: "private",
    operator: "is_empty",
  };

  // if no documentIds are provided, only retrieve information from public documents
  if (!documentIds.length)
    return {
      filters: [publicDocumentsFilter],
    } as CloudRetrieveParams["filters"];

  const privateDocumentsFilter: MetadataFilter = {
    key: "file_id", // Note: LLamaCloud uses "file_id" to reference private document ids as "doc_id" is a restricted field in LlamaCloud
    value: documentIds,
    operator: "in",
  };

  // if documentIds are provided, retrieve information from public and private documents
  return {
    filters: [publicDocumentsFilter, privateDocumentsFilter],
    condition: "or",
  } as CloudRetrieveParams["filters"];
}
